File: web/src/lib/sync/parsers.ts
import fs from "fs";
import readline from "readline";
import type {
  ClaudeMessage,
  ClaudeAssistantMessage,
  ClaudeUserMessage,
  ContentBlock,
} from "@/types/claude";

export interface ParsedMessage {
  uuid: string | null;
  parentUuid: string | null;
  type: string;
  role: string | null;
  isSidechain: boolean;
  agentId: string | null;
  model: string | null;
  contentText: string;
  contentJson: string | null;
  hasToolUse: boolean;
  hasThinking: boolean;
  toolNames: string[];
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
  stopReason: string | null;
  timestamp: string | null;
  lineNumber: number;
}

export interface SessionParseResult {
  messages: ParsedMessage[];
  metadata: {
    firstPrompt: string | null;
    customTitle: string | null;
    models: Map<string, number>;
    totalInputTokens: number;
    totalOutputTokens: number;
    totalCacheCreationTokens: number;
    totalCacheReadTokens: number;
    toolUseCount: number;
    userMessageCount: number;
    assistantMessageCount: number;
    gitBranch: string | null;
    cwd: string | null;
    version: string | null;
    permissionMode: string | null;
    createdAt: string | null;
    modifiedAt: string | null;
  };
}

function extractContentText(content: string | ContentBlock[]): string {
  if (typeof content === "string") return content;
  if (!Array.isArray(content)) return "";

  const parts: string[] = [];
  for (const block of content) {
    if (block.type === "text") {
      parts.push(block.text);
    } else if (block.type === "tool_result") {
      if (
        typeof block.content === "string" &&
        block.content.length < 1000 &&
        !block.is_error
      ) {
        parts.push(block.content);
      } else if (block.is_error && typeof block.content === "string") {
        parts.push(`[Error: ${block.content.slice(0, 200)}]`);
      }
    } else if (block.type === "tool_use") {
      parts.push(`[Tool: ${block.name}]`);
    }
    // Skip thinking blocks from content_text
  }
  return parts.join("\n");
}

function extractAssistantInfo(msg: ClaudeAssistantMessage) {
  const content = msg.message?.content;
  if (!Array.isArray(content))
    return {
      contentText: "",
      hasToolUse: false,
      hasThinking: false,
      toolNames: [] as string[],
    };

  const textParts: string[] = [];
  const toolNames: string[] = [];
  let hasToolUse = false;
  let hasThinking = false;

  for (const block of content) {
    if (block.type === "text") {
      textParts.push(block.text);
    } else if (block.type === "tool_use") {
      hasToolUse = true;
      toolNames.push(block.name);
    } else if (block.type === "thinking") {
      hasThinking = true;
    }
  }

  return { contentText: textParts.join("\n"), hasToolUse, hasThinking, toolNames };
}

function isStreamingPartial(msg: ClaudeAssistantMessage): boolean {
  return (
    msg.message?.stop_reason === null &&
    msg.message?.usage?.inference_geo === "not_available"
  );
}

export async function parseSessionFile(
  filePath: string
): Promise<SessionParseResult> {
  const messages: ParsedMessage[] = [];
  const metadata: SessionParseResult["metadata"] = {
    firstPrompt: null,
    customTitle: null,
    models: new Map(),
    totalInputTokens: 0,
    totalOutputTokens: 0,
    totalCacheCreationTokens: 0,
    totalCacheReadTokens: 0,
    toolUseCount: 0,
    userMessageCount: 0,
    assistantMessageCount: 0,
    gitBranch: null,
    cwd: null,
    version: null,
    permissionMode: null,
    createdAt: null,
    modifiedAt: null,
  };

  // Track assistant message.id for deduplication
  const assistantMessageIds = new Map<string, number>(); // message.id -> index in messages[]
  let lineNumber = 0;

  const fileStream = fs.createReadStream(filePath, { encoding: "utf-8" });
  const rl = readline.createInterface({ input: fileStream, crlfDelay: Infinity });

  for await (const line of rl) {
    lineNumber++;
    if (!line.trim()) continue;

    let parsed: ClaudeMessage;
    try {
      parsed = JSON.parse(line);
    } catch {
      continue; // Skip malformed lines
    }

    // Skip types we don't store
    if (
      parsed.type === "progress" ||
      parsed.type === "file-history-snapshot" ||
      parsed.type === "last-prompt"
    ) {
      continue;
    }

    // Handle custom-title
    if (parsed.type === "custom-title") {
      metadata.customTitle = parsed.customTitle;
      continue;
    }

    // Handle agent-name as fallback title
    if (parsed.type === "agent-name") {
      if (!metadata.customTitle) {
        metadata.customTitle = parsed.agentName;
      }
      continue;
    }

    // Handle system messages (extract duration)
    if (parsed.type === "system") {
      continue;
    }

    // Handle user messages
    if (parsed.type === "user") {
      const userMsg = parsed as ClaudeUserMessage;

      // Skip meta/command messages
      if (userMsg.isMeta) continue;

      const contentText = extractContentText(userMsg.message?.content);

      // Track first user message that isn't a tool result
      if (
        !metadata.firstPrompt &&
        typeof userMsg.message?.content === "string" &&
        !userMsg.isSidechain
      ) {
        metadata.firstPrompt = userMsg.message.content.slice(0, 500);
      }

      // Track metadata from first non-sidechain user message
      if (!userMsg.isSidechain) {
        if (!metadata.gitBranch && userMsg.gitBranch) {
          metadata.gitBranch = userMsg.gitBranch;
        }
        if (!metadata.cwd && userMsg.cwd) {
          metadata.cwd = userMsg.cwd;
        }
        if (!metadata.version && userMsg.version) {
          metadata.version = userMsg.version;
        }
        if (!metadata.permissionMode && userMsg.permissionMode) {
          metadata.permissionMode = userMsg.permissionMode;
        }
      }

      // Track timestamps
      if (userMsg.timestamp) {
        if (!metadata.createdAt || userMsg.timestamp < metadata.createdAt) {
          metadata.createdAt = userMsg.timestamp;
        }
        if (!metadata.modifiedAt || userMsg.timestamp > metadata.modifiedAt) {
          metadata.modifiedAt = userMsg.timestamp;
        }
      }

      // Determine role based on content type
      const isToolResult = Array.isArray(userMsg.message?.content);

      if (!userMsg.isSidechain && !isToolResult) {
        metadata.userMessageCount++;
      }

      messages.push({
        uuid: userMsg.uuid || null,
        parentUuid: userMsg.parentUuid || null,
        type: "user",
        role: "user",
        isSidechain: userMsg.isSidechain,
        agentId: userMsg.agentId || null,
        model: null,
        contentText,
        contentJson: isToolResult
          ? JSON.stringify(userMsg.message.content)
          : null,
        hasToolUse: false,
        hasThinking: false,
        toolNames: [],
        inputTokens: 0,
        outputTokens: 0,
        cacheCreationTokens: 0,
        cacheReadTokens: 0,
        stopReason: null,
        timestamp: userMsg.timestamp || null,
        lineNumber,
      });
      continue;
    }

    // Handle assistant messages
    if (parsed.type === "assistant") {
      const assistantMsg = parsed as ClaudeAssistantMessage;
      const msgId = assistantMsg.message?.id;

      // Deduplication: skip streaming partials
      if (isStreamingPartial(assistantMsg) && msgId) {
        // Store index in case final message comes later
        // We'll add it now but may replace it
        const existingIdx = assistantMessageIds.get(msgId);
        if (existingIdx !== undefined) continue; // Already have a better version
      }

      const { contentText, hasToolUse, hasThinking, toolNames } =
        extractAssistantInfo(assistantMsg);

      const usage = assistantMsg.message?.usage;
      const inputTokens = usage?.input_tokens || 0;
      const outputTokens = usage?.output_tokens || 0;
      const cacheCreation = usage?.cache_creation_input_tokens || 0;
      const cacheRead = usage?.cache_read_input_tokens || 0;

      const model = assistantMsg.message?.model || null;
      const stopReason = assistantMsg.message?.stop_reason || null;

      const parsedMsg: ParsedMessage = {
        uuid: assistantMsg.uuid || null,
        parentUuid: assistantMsg.parentUuid || null,
        type: "assistant",
        role: "assistant",
        isSidechain: assistantMsg.isSidechain,
        agentId: assistantMsg.agentId || null,
        model,
        contentText,
        contentJson: JSON.stringify(assistantMsg.message?.content || []),
        hasToolUse,
        hasThinking,
        toolNames,
        inputTokens,
        outputTokens,
        cacheCreationTokens: cacheCreation,
        cacheReadTokens: cacheRead,
        stopReason,
        timestamp: assistantMsg.timestamp || null,
        lineNumber,
      };

      // Dedup: if we already have a message with same id, replace if this one is final
      if (msgId) {
        const existingIdx = assistantMessageIds.get(msgId);
        if (existingIdx !== undefined) {
          // Replace previous entry if this one has stop_reason (is final)
          if (stopReason !== null) {
            messages[existingIdx] = parsedMsg;
          }
          continue;
        }
        assistantMessageIds.set(msgId, messages.length);
      }

      // Track metadata
      if (!assistantMsg.isSidechain) {
        metadata.assistantMessageCount++;
        if (hasToolUse) metadata.toolUseCount += toolNames.length;
        metadata.totalInputTokens += inputTokens;
        metadata.totalOutputTokens += outputTokens;
        metadata.totalCacheCreationTokens += cacheCreation;
        metadata.totalCacheReadTokens += cacheRead;

        if (model) {
          metadata.models.set(model, (metadata.models.get(model) || 0) + 1);
        }
      }

      if (assistantMsg.timestamp) {
        if (
          !metadata.modifiedAt ||
          assistantMsg.timestamp > metadata.modifiedAt
        ) {
          metadata.modifiedAt = assistantMsg.timestamp;
        }
      }

      messages.push(parsedMsg);
    }
  }

  return { messages, metadata };
}

export function getPrimaryModel(models: Map<string, number>): string | null {
  if (models.size === 0) return null;
  let maxCount = 0;
  let primary: string | null = null;
  for (const [model, count] of models) {
    if (count > maxCount) {
      maxCount = count;
      primary = model;
    }
  }
  return primary;
}

File: src-tauri/src/sync/parsers.rs
use std::collections::HashMap;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::path::Path;

use crate::types::claude::{AssistantMessageInner, ContentBlock, GenericMessage};

#[derive(Debug, Clone)]
pub struct ParsedMessage {
    pub uuid: Option<String>,
    pub parent_uuid: Option<String>,
    pub msg_type: String,
    pub role: Option<String>,
    pub is_sidechain: bool,
    pub agent_id: Option<String>,
    pub model: Option<String>,
    pub content_text: String,
    pub content_json: Option<String>,
    pub has_tool_use: bool,
    pub has_thinking: bool,
    pub tool_names: Vec<String>,
    pub input_tokens: i64,
    pub output_tokens: i64,
    pub stop_reason: Option<String>,
    pub timestamp: Option<String>,
    pub line_number: i64,
}

#[derive(Debug)]
pub struct SessionMetadata {
    pub first_prompt: Option<String>,
    pub custom_title: Option<String>,
    pub models: HashMap<String, i64>,
    pub total_input_tokens: i64,
    pub total_output_tokens: i64,
    pub total_cache_creation_tokens: i64,
    pub total_cache_read_tokens: i64,
    pub tool_use_count: i64,
    pub user_message_count: i64,
    pub assistant_message_count: i64,
    pub git_branch: Option<String>,
    pub cwd: Option<String>,
    pub version: Option<String>,
    pub permission_mode: Option<String>,
    pub created_at: Option<String>,
    pub modified_at: Option<String>,
}

pub struct SessionParseResult {
    pub messages: Vec<ParsedMessage>,
    pub metadata: SessionMetadata,
}

fn extract_content_text_from_value(content: &serde_json::Value) -> String {
    if let Some(s) = content.as_str() {
        return s.to_string();
    }
    if let Some(arr) = content.as_array() {
        let mut parts = Vec::new();
        for block in arr {
            let block_type = block.get("type").and_then(|t| t.as_str()).unwrap_or("");
            match block_type {
                "text" => {
                    if let Some(text) = block.get("text").and_then(|t| t.as_str()) {
                        parts.push(text.to_string());
                    }
                }
                "tool_result" => {
                    let is_error = block
                        .get("is_error")
                        .and_then(|e| e.as_bool())
                        .unwrap_or(false);
                    if let Some(c) = block.get("content").and_then(|c| c.as_str()) {
                        if is_error {
                            parts.push(format!("[Error: {}]", &c[..c.len().min(200)]));
                        } else if c.len() < 1000 {
                            parts.push(c.to_string());
                        }
                    }
                }
                "tool_use" => {
                    if let Some(name) = block.get("name").and_then(|n| n.as_str()) {
                        parts.push(format!("[Tool: {}]", name));
                    }
                }
                _ => {}
            }
        }
        return parts.join("\n");
    }
    String::new()
}

fn extract_assistant_info(
    content: &[ContentBlock],
) -> (String, bool, bool, Vec<String>) {
    let mut text_parts = Vec::new();
    let mut tool_names = Vec::new();
    let mut has_tool_use = false;
    let mut has_thinking = false;

    for block in content {
        match block {
            ContentBlock::Text { text } => {
                text_parts.push(text.clone());
            }
            ContentBlock::ToolUse { name, .. } => {
                has_tool_use = true;
                tool_names.push(name.clone());
            }
            ContentBlock::Thinking { .. } => {
                has_thinking = true;
            }
            ContentBlock::ToolResult { .. } => {}
        }
    }

    (text_parts.join("\n"), has_tool_use, has_thinking, tool_names)
}

fn is_streaming_partial(inner: &AssistantMessageInner) -> bool {
    inner.stop_reason.is_none()
        && inner
            .usage
            .as_ref()
            .and_then(|u| u.inference_geo.as_deref())
            == Some("not_available")
}

pub fn parse_session_file(file_path: &Path) -> Result<SessionParseResult, String> {
    let file =
        File::open(file_path).map_err(|e| format!("Failed to open {}: {}", file_path.display(), e))?;
    let reader = BufReader::new(file);

    let mut messages: Vec<ParsedMessage> = Vec::new();
    let mut metadata = SessionMetadata {
        first_prompt: None,
        custom_title: None,
        models: HashMap::new(),
        total_input_tokens: 0,
        total_output_tokens: 0,
        total_cache_creation_tokens: 0,
        total_cache_read_tokens: 0,
        tool_use_count: 0,
        user_message_count: 0,
        assistant_message_count: 0,
        git_branch: None,
        cwd: None,
        version: None,
        permission_mode: None,
        created_at: None,
        modified_at: None,
    };

    let mut assistant_message_ids: HashMap<String, usize> = HashMap::new();
    let mut line_number: i64 = 0;

    for line_result in reader.lines() {
        line_number += 1;
        let line = match line_result {
            Ok(l) => l,
            Err(_) => continue,
        };
        let trimmed = line.trim();
        if trimmed.is_empty() {
            continue;
        }

        let parsed: GenericMessage = match serde_json::from_str(trimmed) {
            Ok(m) => m,
            Err(_) => continue,
        };

        let msg_type = parsed.msg_type.as_str();

        // Skip types we don't store
        if matches!(
            msg_type,
            "progress" | "file-history-snapshot" | "last-prompt" | "system"
        ) {
            continue;
        }

        // Handle custom-title
        if msg_type == "custom-title" {
            if let Some(title) = parsed.custom_title {
                metadata.custom_title = Some(title);
            }
            continue;
        }

        // Handle agent-name
        if msg_type == "agent-name" {
            if metadata.custom_title.is_none() {
                if let Some(name) = parsed.agent_name {
                    metadata.custom_title = Some(name);
                }
            }
            continue;
        }

        let is_sidechain = parsed.is_sidechain.unwrap_or(false);

        // Handle user messages
        if msg_type == "user" {
            if parsed.is_meta.unwrap_or(false) {
                continue;
            }

            let message_val = parsed.message.as_ref();
            let content_val = message_val.and_then(|m| m.get("content"));

            let content_text = content_val
                .map(|c| extract_content_text_from_value(c))
                .unwrap_or_default();

            let is_tool_result = content_val
                .map(|c| c.is_array())
                .unwrap_or(false);

            // Track first prompt
            if metadata.first_prompt.is_none()
                && !is_sidechain
                && content_val.map(|c| c.is_string()).unwrap_or(false)
            {
                let prompt = content_val
                    .and_then(|c| c.as_str())
                    .unwrap_or("")
                    .chars()
                    .take(500)
                    .collect::<String>();
                metadata.first_prompt = Some(prompt);
            }

            // Track metadata from non-sidechain messages
            if !is_sidechain {
                if metadata.git_branch.is_none() {
                    if let Some(ref b) = parsed.git_branch {
                        if !b.is_empty() {
                            metadata.git_branch = Some(b.clone());
                        }
                    }
                }
                if metadata.cwd.is_none() {
                    metadata.cwd = parsed.cwd.clone();
                }
                if metadata.version.is_none() {
                    metadata.version = parsed.version.clone();
                }
                if metadata.permission_mode.is_none() {
                    metadata.permission_mode = parsed.permission_mode.clone();
                }
            }

            // Track timestamps
            if let Some(ref ts) = parsed.timestamp {
                if metadata
                    .created_at
                    .as_ref()
                    .map_or(true, |c| ts < c)
                {
                    metadata.created_at = Some(ts.clone());
                }
                if metadata
                    .modified_at
                    .as_ref()
                    .map_or(true, |m| ts > m)
                {
                    metadata.modified_at = Some(ts.clone());
                }
            }

            if !is_sidechain && !is_tool_result {
                metadata.user_message_count += 1;
            }

            let content_json = if is_tool_result {
                content_val.map(|c| serde_json::to_string(c).unwrap_or_default())
            } else {
                None
            };

            messages.push(ParsedMessage {
                uuid: parsed.uuid,
                parent_uuid: parsed.parent_uuid,
                msg_type: "user".to_string(),
                role: Some("user".to_string()),
                is_sidechain,
                agent_id: parsed.agent_id,
                model: None,
                content_text,
                content_json,
                has_tool_use: false,
                has_thinking: false,
                tool_names: Vec::new(),
                input_tokens: 0,
                output_tokens: 0,
                stop_reason: None,
                timestamp: parsed.timestamp,
                line_number,
            });
            continue;
        }

        // Handle assistant messages
        if msg_type == "assistant" {
            let inner: AssistantMessageInner = match parsed
                .message
                .as_ref()
                .and_then(|m| serde_json::from_value(m.clone()).ok())
            {
                Some(inner) => inner,
                None => continue,
            };

            let msg_id = inner.id.clone();

            // Skip streaming partials if we already have one
            if is_streaming_partial(&inner) {
                if let Some(ref id) = msg_id {
                    if assistant_message_ids.contains_key(id) {
                        continue;
                    }
                }
            }

            let content_blocks = inner.content.as_deref().unwrap_or(&[]);
            let (content_text, has_tool_use, has_thinking, tool_names) =
                extract_assistant_info(content_blocks);

            let usage = inner.usage.as_ref();
            let input_tokens = usage.and_then(|u| u.input_tokens).unwrap_or(0);
            let output_tokens = usage.and_then(|u| u.output_tokens).unwrap_or(0);
            let cache_creation = usage
                .and_then(|u| u.cache_creation_input_tokens)
                .unwrap_or(0);
            let cache_read = usage
                .and_then(|u| u.cache_read_input_tokens)
                .unwrap_or(0);

            let model = inner.model.clone();
            let stop_reason = inner.stop_reason.clone();

            let content_json = serde_json::to_string(content_blocks).ok();

            let parsed_msg = ParsedMessage {
                uuid: parsed.uuid,
                parent_uuid: parsed.parent_uuid,
                msg_type: "assistant".to_string(),
                role: Some("assistant".to_string()),
                is_sidechain,
                agent_id: parsed.agent_id,
                model: model.clone(),
                content_text,
                content_json,
                has_tool_use,
                has_thinking,
                tool_names: tool_names.clone(),
                input_tokens,
                output_tokens,
                stop_reason: stop_reason.clone(),
                timestamp: parsed.timestamp.clone(),
                line_number,
            };

            // Dedup: if we already have a message with same id, replace if this one is final
            if let Some(ref id) = msg_id {
                if let Some(&existing_idx) = assistant_message_ids.get(id) {
                    if stop_reason.is_some() {
                        messages[existing_idx] = parsed_msg;
                    }
                    continue;
                }
                assistant_message_ids.insert(id.clone(), messages.len());
            }

            // Track metadata
            if !is_sidechain {
                metadata.assistant_message_count += 1;
                if has_tool_use {
                    metadata.tool_use_count += tool_names.len() as i64;
                }
                metadata.total_input_tokens += input_tokens;
                metadata.total_output_tokens += output_tokens;
                metadata.total_cache_creation_tokens += cache_creation;
                metadata.total_cache_read_tokens += cache_read;

                if let Some(ref m) = model {
                    *metadata.models.entry(m.clone()).or_insert(0) += 1;
                }
            }

            if let Some(ref ts) = parsed.timestamp {
                if metadata
                    .modified_at
                    .as_ref()
                    .map_or(true, |m| ts > m)
                {
                    metadata.modified_at = Some(ts.clone());
                }
            }

            messages.push(parsed_msg);
        }
    }

    Ok(SessionParseResult { messages, metadata })
}

pub fn get_primary_model(models: &HashMap<String, i64>) -> Option<String> {
    models
        .iter()
        .max_by_key(|(_, count)| *count)
        .map(|(model, _)| model.clone())
}
